components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java

```diff
@@ -0,0 +1,248 @@
+  //
+
+  private static String getUnixSocketFolderOrAbort() {
+    // Environment variable "parsing"
+    final String unixSocketPath = getEnvVarOrDefault(
+        Constants.EnvironmentVariable.DAPR_COMPONENTS_SOCKETS_PATH,
+        Constants.Defaults.DAPR_COMPONENTS_SOCKETS_PATH);
+    // Validate directory exists and is a writable folder
+    final File targetFolder = new File(unixSocketPath);
+    final boolean isValidPath = targetFolder.exists()
+        && targetFolder.isDirectory()
+        && targetFolder.canWrite();
+    if (!isValidPath) {
+      final String errMsg = "ERROR: path '" + unixSocketPath + "' is not an existing writable directory. Aborting";
+      logFatalAndAbort(errMsg);
+    }
+
+    log.info("Unix Domain Socket files will be created on " + targetFolder.toPath().toAbsolutePath());
+
+    return unixSocketPath;
+  }
+
+  // While this does what its name says, it is not really testable.
+  // We should consider if migrating this to exceptions would be better.
+  // Created components-java-sdk#21 to track it.
+  private static void logFatalAndAbort(String errMsg) {
+    log.severe(errMsg);
+    System.err.println(errMsg);
+    System.exit(1);
+  }
+
+  private static String getEnvVarOrDefault(final String envVarName, final String defaultValue) {
+    String value = System.getenv(envVarName);
+    if (value == null) {
+      log.info("Environment variable " + envVarName + " is not defined. "
+          + " Assuming default value of " + defaultValue);
+      return defaultValue;
+    }
+    return value;
+  }
+
+  private Path buildPathForComponentUnixDomainSocket(final String componentName) {
+    final String suffix = getEnvVarOrDefault(
+        Constants.EnvironmentVariable.DAPR_COMPONENTS_SOCKET_EXTENSION,
+        Constants.Defaults.DAPR_COMPONENTS_SOCKET_EXTENSION);
+    final String udsFilename = componentName + '.' + suffix;
+    final Path targetFolder = Paths.get(unixDomainSocketFolder);
+    return targetFolder.resolve(udsFilename);
+  }
+
+
+  //
+  // Server lifecycle management methods
+  //
+
+
+  private void startServers() throws IOException {
+    // For each registered server, start it and set up add it to a compound shutdown hook.
+    for (var componentNameAndServer : servers.entrySet()) {
+      final var componentName = componentNameAndServer.getKey();
+      final var server = componentNameAndServer.getValue();
+
+      server.start();
+      log.info("Started server for " + componentName);
+    }
+
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+      // Use stderr here since the logger may have been reset by its JVM shutdown hook.
+      System.err.println("*** shutting down gRPC server since JVM is shutting down");
+      servers.forEach((componentName, server) -> {
+        try {
+          System.err.println("    *** shutting doing server instance " + server.toString());
+          server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
```

Comment:
Questions as one not familiar with the Java gRPC library: would the servers already be shutting down at this point, and that's just being awaited here, or is this initiating the shutdown?  Is this wait sequential?  Would it add up in the case of multiple servers (vs. the parallel await in the `blockUntilShutdown()` method.

Reply:
> Questions as one not familiar with the Java gRPC library: 
> would the servers already be shutting down at this point, and that's just being awaited here, or is this initiating the shutdown?

This code sequence initialize a "clean" shutdown if the server process receives a signal to terminal (Say, a Ctrl-C/ sig KILL).

>  Is this wait sequential? Would it add up in the case of multiple servers (vs. the parallel await in the `blockUntilShutdown()` method.

 Yes, it is sequential. I can make it parallel.